subsys/bluetooth/mesh/transport.c

```diff
@@ -1127,18 +1134,52 @@ static void seg_rx_reset(struct seg_rx *rx, bool full_reset)
 	}
 }
 
-static void seg_ack(struct k_work *work)
+static void seg_handler(struct k_work *work)
 {
-	struct seg_rx *rx = CONTAINER_OF(work, struct seg_rx, ack);
+	NET_BUF_SIMPLE_DEFINE(seg_buf, BT_MESH_RX_SDU_MAX);
+	struct seg_rx *rx = CONTAINER_OF(work, struct seg_rx, work);
 	int32_t timeout;
 
-	if (!rx->in_use || rx->block == BLOCK_COMPLETE(rx->seg_n)) {
+	if (!rx->in_use) {
 		/* Cancellation of this timer may have failed. If it fails as
 		 * part of seg_reset, in_use will be false.
 		 * If it fails as part of the processing of a fully received
 		 * SDU, the ack is already being sent from the receive handler,
 		 * and the timer based ack sending can be ignored.
 		 */
+
+		return;
+	} else if (rx->block == BLOCK_COMPLETE(rx->seg_n)) {
+		struct net_buf_simple sdu;
+		struct bt_mesh_net_rx net_rx = {
+			.sub = rx->sub,
+			.ctx = {
+				.net_idx = rx->sub->net_idx,
+				.addr = rx->src,
+				.recv_dst = rx->dst,
+				.recv_ttl = rx->ttl,
+				.recv_rssi = rx->recv_rssi,
+			},
+			.net_if = rx->net_if,
+			.old_iv = (uint32_t)(rx->seq_auth >> 24) != bt_mesh.iv_index,
+			.new_key = rx->new_key,
+			.friend_cred = rx->friend_cred,
+			.seq = rx->seq_auth & BIT_MASK(24),
+			.local_match = !(rx->obo),
+			.friend_match = rx->obo,
+		};
+
+		/* Decrypting in place to avoid creating two assembly buffers.
+		 * We'll reassemble the buffer from the segments before each
+		 * decryption attempt.
+		 */
+		net_buf_simple_init(&seg_buf, 0);
+		net_buf_simple_init_with_data(
+			&sdu, seg_buf.data, rx->len - APP_MIC_LEN(rx->szmic));
+
+		(void)sdu_recv(&net_rx, rx->hdr, rx->szmic, &seg_buf, &sdu, rx);
```

Comment:
This can cause unintended out of order reception if we receive an unsegmented message from the same device right after the last segment, as the workqueue won't be processed until after. Normally, out of order packets are rejected by the replay protection, but as that  happens in trans_seg, it won't save us.

For instance, imagine if a device sends a one-segment segmented message to open a door, then sends a second unsegmented message to lock all the doors. Before, at worst, the second message would be ignored if it manages to race past the first one, but now, it may get accepted and processed while the segmented message sits in the workqueue, and there's no way for the app to detect or prevent this. 

Could we prevent this somehow?

Reply:
I don't think this is a problem. First of all, mesh segment message can't ensure that it can be accurately received by the other party at one time, even if it has only one segment message.
The correct approach should be to send another message after receiving the reply, if these messages are important.